File: 636. Exclusive Time of Functions.py
class Solution:
    def exclusiveTime(self, n: int, logs: List[str]) -> List[int]:
        '''
        Using stack to keep track of the logs
        ''' 
        ans = [0 for _ in range(n)] # ans array
        stack = []                  # the stack for keep tracking
        prev = 0                    # the var for recording prev timestamp
        for log in logs:
            log = log.split(':')
            # for start condition
            if log[1] == 'start':
                # if there is a switch of functions
                if stack:
                    # update the being switched function
                    ans[stack[-1]] += int(log[2]) - prev
                # put the new one into stack and update the timestamp
                stack.append(int(log[0]))
                prev = int(log[2])
            # for end condition
            else:
                # there must be a function in stack, update its time
                ans[stack[-1]] += int(log[2]) - prev + 1
                stack.pop()
                prev = int(log[2]) + 1
        return ans

File: 636. Exclusive Time of Functions.java
class Solution {
    public int[] exclusiveTime(int n, List<String> logs) {
        int[] ans = new int[n];
        Stack<Integer> s = new Stack<>();
        int prev = 0;
        for (String log : logs) {
            String[] l = log.split(":");
            if (l[1].equals("start")) {
                if (!s.empty()) {
                    ans[s.peek()] += Integer.parseInt(l[2]) - prev;
                }
                s.push(Integer.parseInt(l[0]));
                prev = Integer.parseInt(l[2]);
            } else {
                ans[s.peek()] += Integer.parseInt(l[2]) - prev + 1;
                s.pop();
                prev = Integer.parseInt(l[2])+1;
            }
        }
        return ans;
    }
}